### DISCUSSION TITLE
Long Query Process Times, Looking for Suggestions How to Resolve or Define

### DISCUSSION THREAD

**Discussion by dmelideo** [AdguardTeam/AdGuardHome#6867] [Q&A]
I am trying to understand why my AdGuard home has recently exhibited a significant increase in Average Processing Time. I used to about 40ms, then jumped to about 70ms and am now at 165ms. I have some "processed" queries between 200ms and 700ms and a few at greater than 1000ms with one at just over 10000ms. All appear to be ring or amazon devices, cameras and echo's. 

Is there an easy way to export the query log so I can better evaluate the offenders? Currently just scrolling down thru the list to identify interesting occurrences.

Would be great to have a way to on the dashboard to see top "x" long processing time devices (suggestion)

Others have reported getting annoying adds thru the amazon ecosystem, we are not getting any which means blocking is working. 
Asking for thoughts on how to possibly correct these occurrences by device (IP address). I am only using 5 DNS blocklists.

**Comment by ghost** [comment]
Feel free to add your suggestions to the issue tracker as a feature request.

As for analysing the increase in query latency, you should find them within the AdGuard Home log directory, depending on how you're running the AGH server.

Once you've found them, you run the command `cat data/querylog.json | jq` to analyse it a bit better.

You could also export it to csv/xlsx for easier analysis with filtering with 
`cat data/querylog.json | jq -r '[.time, .client, .query, .type] | @csv' > querylogs.csv`

**Reply by dmelideo** [reply]
Thank you

**Comment by Dynamic5912** [comment]
> I am trying to understand why my AdGuard home has recently exhibited a significant increase in Average Processing Time. I used to about 40ms, then jumped to about 70ms and am now at 165ms. I have some "processed" queries between 200ms and 700ms and a few at greater than 1000ms with one at just over 10000ms. All appear to be ring or amazon devices, cameras and echo's. 
> 
> 
> 
> Is there an easy way to export the query log so I can better evaluate the offenders? Currently just scrolling down thru the list to identify interesting occurrences.
> 
> 
> 
> Would be great to have a way to on the dashboard to see top "x" long processing time devices (suggestion)
> 
> 
> 
> Others have reported getting annoying adds thru the amazon ecosystem, we are not getting any which means blocking is working. 
> 
> Asking for thoughts on how to possibly correct these occurrences by device (IP address). I am only using 5 DNS blocklists.

Do you have caching enabled?

Do you have Parental Control and/or Malware blocking enabled?

**Comment by andylau004** [comment]
@dmelideo 
just like this question  #6901  ???

advise change  AdGuardHome.yaml config (response time more faster,, less cost)

new configuration items below ,,,,
```
ratelimit: 0
cache_size: 16777220
  upstream_dns:
    - 94.140.14.140:53
    - 223.5.5.5:53
upstream_timeout: 2s
max_goroutines: 500
upstream_mode: load_balance
```

may be `upstream_mode: parallel`      will more effective then load_balance

**Comment by hokfujow** [comment]
I've had the exact same issue and it started around the same time.   I think I just solved it for my case.  Previously my processing times were under 10ms.  After an update, processing time on two different instances increased by 10x to 100x.  I tried a lot of things, but the thing that I believe got me back to previous experience (under 10ms) was putting my router address into the private reverse DNS server field.  I had always left this field blank because I didn't know what to put there.  I had the first reverse DNS box checked, and adguard showed the server that was being used by default.  My standalone AGH instance had been showing the address of my Home Assistant AGH address a the default, while the HA OS AGH instance was showing its internal address as the default. 
Putting the router address into the field on both instances immediately restored performance. 
The thing is I had always had this field blank.  Possibly something changed in how the the private reverse DNS works such that leaving the field blank caused slow performance whereas it didn't before.  I couldn't find anything in release notes that was an obvious explanation but  I'm quite confident that this change made the difference - for me at least.

**Reply by spicymanp** [reply]
I just wanted to say thank you for bringing this to my attention. I was having issues with response times being over 10000ms and going in and adding my router address to the Private Reverse DNS Server field and checking the little tick box under it has seemingly fixed the issue.

**Reply by Dynamic5912** [reply]
This has nothing to do with DNS resolution times

**Reply by spicymanp** [reply]
oh :(

**Reply by hokfujow** [reply]
Actually, now I seem to be unable to add my router address to private reverse DNS field at all.  I had it in at one time and took it out during troubleshooting - now it won't accept it all.  Very strange.  The most recen thing I did to address this was changed some firewall rules related to DNS port that I think might have been slowing things down.  I'm not 100% sure that was the reason. I also removed a couple filtering lists in Adguard.  At the moment, it's working very well (times down to 3 ms) , so I'm not messing with it.

**Reply by cornerwilderness** [reply]
Hi, I had the same issue also. My Average processing times were something like 5000ms and DNS was generally not working at all.
I did the above and added my router to my reverse DNS but as I used the onboard testing tool (test upstreams) I noticed that my Upstream DNS servers only had one line and it wasn't working. 
I think this might have been reset at some point automatically, because I started having these issues around last Christmas.

Anyhow this is how I fixed it:

Settings -> DNS settings

Upstream DNS servers -> I added my service providers DNS addresses here, you can choose whatever you like, there is even a link to providers on the description.

Private reverse DNS servers:
Your router IP

--> Test upstreams : to see if everything is OK
